Add log download

wsprent opened the pull request:
Added two links to the bottom bar of the logs which generate plaintext versions of the log with and without headers respectively.

Currently it does so in new tabs which could be changed by altering the "content-disposition" header to "attachment", but with copy/pasting currently disregarding newlines I think it's alright.

williams-unity committed:
Added children /text and /text_with_headers to /logs/$logname reachable from the bottom-bar of the log page.
Modified logs.py to respond with plaintext logs when they are invoked.

williams-unity committed:
resolved merge conflict from cherrypicking

williams-unity committed:
resolved merge conflict from cherrypicking

williams-unity committed:
Added children /text and /text_with_headers to /logs/$logname reachable from the bottom-bar of the log page.
Modified logs.py to respond with plaintext logs when they are invoked.

williams-unity committed:
resolved merge conflict from cherrypicking

williams-unity committed:
Made path recognition more stringent and reversed if to remove odd indent.

williams-unity committed:
Made path recognition more stringent and reversed if to remove odd indent.

williams-unity committed:
resolved merge conflict from cherrypicking

williams-unity committed:
Added default filenames for logs.

williams-unity committed:
Added default filenames for logs.

williams-unity committed:
Added default filenames for logs.

williams-unity committed:
Added default filenames for logs.

williams-unity committed:
Merge branch 'add-log-download' of https://github.com/wsprent/katana into add-log-download

kkateq commented on master/buildbot/status/web/logs.py:
```diff
@@ -117,41 +122,53 @@ def render_GET(self, req):
+        if self.asText:
+            with_headers = "_with_headers" if self.withHeaders else ""
+            base_name = self.original.step.getName() + "_" + self.original.getName() + with_headers
+            base_name = base_name.replace(" ", "_") + ".txt"
```
It's not necessary to replace spaces as long as you surrounded file name into quotes.

kkateq replied:
Also maybe better to use `.log` extension.

wsprent replied:
The spaces come from the step.getName() call. I think it'll be annoying for the user if we don't remove them.

kkateq replied:
There are some steps names that make me wonder about file name result, e.g. 
`Download Artifact for 'proj0-Build MacEditor'`
`Trigger Dependencies (Trunk or 5.x-Based Branch)`

wsprent replied:
They should turn out as follows. It's a bit garbled, but better than having to \ whitespace in the terminal, I think.

```
   Download_Artifact_for_'proj0-Build_MacEditor'
   Trigger_Dependencies_(Trunk_or_5.x-Based_Branch)
```

I don't have any better ideas at least.

kkateq replied:
Right, but this also doesn't look good. So it's up to you, I would probably try to replace special characters such as `~ ! @ # $ % ^ & * ( )  ; < > ? , [ ] { } ' "` with underscore too.

campos-ddc replied:
Watch out for forward slashes such as step 7 in http://katana.hq.unity3d.com/projects/Unity/builders/proj0-Build%20UserDocumentation%20-%20English%20%28w%2F%20PS3%29/builds/3066?unity_branch=5.4/staging

"Build UserDocumentation - English (w/ PS3)"

campos-ddc replied:
I'd recommend: `re.sub(r'[^\w]', '_', step_name)` to replace all non-alphanumeric with underscores.

wsprent replied:
Yeah, you're both right. Pushed change.

kkateq commented on master/buildbot/status/web/logs.py:
```diff
@@ -117,41 +122,53 @@ def render_GET(self, req):
+            with_headers = "_with_headers" if self.withHeaders else ""
+            base_name = self.original.step.getName() + "_" + self.original.getName() + with_headers
+            base_name = base_name.replace(" ", "_") + ".txt"
+            req.setHeader("Content-Disposition", "inline; filename =\"" + base_name + "\"")
```
Should  `attachment` be here instead of `inline` ?

wsprent replied:
It could be, but since this is a stopgap measure anyway, then having it as  `inline` allows for opening the text file in the browser and copy/pasting or what not.

wsprent replied:
This means that they have to right click to download, but that's a trade off I guess.

kkateq replied:
Is this feature about 'download log' or 'open another tab with the log text and copy/paste' ? It's almost the same as  'Open in new window' isn't it?

wsprent replied:
Yes, but as long as copy/pasting from the logs doesn't work as intended in Chrome, then I think it makes sense to be able to access the raw file in the browser especially since this is a temporary thing. But if you don't think that's a good idea I'll change it.

kkateq replied:
So I would replace 'Open in new Window' with the opening log file in the new tab - the way you did it for download which would be much useful for copying/pasting, and 'Download log' link should download log. Otherwise, it's quite confusing.

wsprent replied:
Good idea, I pushed some commits which implement this.

kkateq commented on www/templates/logs.html:
```diff
@@ -19,12 +19,24 @@
                     /
                 </li>
                 <li>
+		  <a href="{{ plaintext_url }}" target="_blank">Download as plaintext </a>
```
Remove `target="_blank"` otherwise each time you click link browser will try to open new tab.

kkateq replied:
Also, I would remove _as plaintext_ as long as this is the only option.

kkateq commented on www/templates/logs.html:
```diff
@@ -19,12 +19,24 @@
+                    /
+                </li>
+		<li>
+		  <a href="{{ plaintext_with_headers_url }}" target="_blank">Download as plaintext with headers</a>
```
Same here.

williams-unity committed:
Changed names of log download links.
Changed file extension to ".log".
Plaintext logs no longer open in new tabs.

williams-unity committed:
Merge branch 'add-log-download' of https://github.com/wsprent/katana into add-log-download

williams-unity committed:
Modified "Open in new window" option to open a new tab with the plaintext log either with or without headers.
Made the download links trigger a download of the plaintext logs.
Renamed some variables to better reflect how they are used.

williams-unity committed:
Remove all non alphanumeric characters from filename
